### PR TITLE
Append subdirectories of device client current working directory to PATH

### PR DESCRIPTION
**Pull request by marcoemorais-aws**
### Motivation
* Managed job templates feature will look for job handlers in `./job` subdirectory of the default device client config directory and device client current working directory.

#### Change summary
* On startup, add both the default device client config directory and the current working directory as well as standard subdirectories (`./job`) to the `PATH` environment variable.
* This is the value of PATH that is set and logged on startup when I run device client on my ubuntu 20 image.
```
/usr/local/sbin:
/usr/local/bin:
/usr/sbin:
/usr/bin:
/sbin:
/bin:
/root/.aws-iot-device-client:           << added to PATH by device client on startup 
/root/.aws-iot-device-client/jobs:      << added to PATH by device client on startup 
/tmp/aws-iot-device-client-clion:       << added to PATH by device client on startup 
/tmp/aws-iot-device-client-clion/jobs   << added to PATH by device client on startup 
```

#### Revision diff summary
If there is more than one revision, please explain what has been changed since the last revision.

### Testing
 **Is your change tested? If not, please justify the reason.**  
 **Please list your testing steps and test results.** 
* Unit tests that exercise the feature by faking out POSIX system calls to get and set environment variables and the current working directory.
* Functional test using local builds of the device client on amazonlinux2 and ubuntu:20.04.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/util/EnvUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        cwd.resize(PATH_MAX + 1, '\0'); // Fallback to hardcoded PATH_MAX.
+    }
+    else
+    {
+        cwd.resize(path_max + 1, '\0');
+    }
+
+    char *pcwd = nullptr;
+    while (!pcwd)
```
Comment (HarshGandhi-AWS) [inline source/util/EnvUtils.cpp:74]: We should avoid having infinite loop. Considering the PATH_MAX value for most Linux OS, lets end this look with maximum 3 retries.

File: source/util/EnvUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    std::ostringstream oss;
+    if (const char *path = os->getenv(PATH_ENVIRONMENT))
+    {
+        oss << path;
+    }
+    if (oss.tellp())
+    {
+        oss << PATH_ENVIRONMENT_SEPARATOR;
+    }
+    oss << cwd.data() // Current working directory.
```
Comment (HarshGandhi-AWS) [inline source/util/EnvUtils.cpp:102]: We should update this logic to store these three absolute paths PATH environment variables. 
```
~/.aws-iot-device-client/
~/.aws-iot-device-client/jobs
./   <<<<---- Current directory absolute path

```

File: source/jobs/JobEngine.cpp [hunk 3]
```diff
@@ -99,18 +99,17 @@ string JobEngine::buildCommand(Optional<string> path, std::string handler, std::
             "Using path {%s} supplied by job document for command execution",
             Util::Sanitize(path.value()).c_str());
         commandStream << path.value();
+        constexpr char separator = '/';
```
[the remark this answers is not in the document]
Comment (marcoemorais-aws) [reply source/jobs/JobEngine.cpp]: Done.

File: source/main.cpp [hunk 4]
```diff
@@ -266,6 +267,15 @@ int main(int argc, char *argv[])
         LoggerFactory::reconfigure(config.config);
     }
 
+    EnvUtils envUtils;
+    if (envUtils.AppendCwdToPath())
+    {
+        // Failure to append current working directory is not a fatal error,
+        // but some features of device client such as standard job execution
```
[the remark this answers is not in the document]
Comment (marcoemorais-aws) [reply source/main.cpp]: Done.

File: source/util/EnvUtils.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                    virtual char *getcwd(char *buf, size_t size) = 0;
+                };
+
+                /**
+                 * \brief OSInterfacePosix is an operating system interface using posix system calls.
+                 */
+                struct OSInterfacePosix : public OSInterface
+                {
+                    /**
+                     * \brief Get get value of an environment variable.
```
[the remark this answers is not in the document]
Comment (marcoemorais-aws) [reply source/util/EnvUtils.h]: Done.

File: source/util/EnvUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    char *pcwd = nullptr;
+    while (!pcwd)
+    {
+        pcwd = os->getcwd(cwd.data(), cwd.size() - 1); // Leave room for terminating null.
+        if (!pcwd)
+        {
+            if (errno == ERANGE)
+            {
+                cwd.resize(cwd.size() * 2, '\0');
```
[the remark this answers is not in the document]
Comment (marcoemorais-aws) [reply source/util/EnvUtils.cpp]: No, std::vector::resize (linked below) is capable of resizing larger (and smaller).

https://en.cppreference.com/w/cpp/container/vector/resize

File: source/util/EnvUtils.cpp (hunk 1 above)
Comment (marcoemorais-aws) [reply source/util/EnvUtils.cpp:81]: Done.  We will resize at most 3 times from initial size. Unit test added to exercise this failure mode.

File: source/util/EnvUtils.cpp (hunk 2 above)
Comment (marcoemorais-aws) [reply source/util/EnvUtils.cpp]: This is the value of PATH that is set and logged on startup when I run device client on my ubuntu 20 image with default config directory `/root/.aws-iot-device-client` and device client invoked from current working directory `/tmp`.    

```
/usr/local/sbin:
/usr/local/bin:
/usr/sbin:
/usr/bin:
/sbin:
/bin:
/root/.aws-iot-device-client:
/root/.aws-iot-device-client/jobs:
/tmp/aws-iot-device-client-clion:
/tmp/aws-iot-device-client-clion/jobs
```

I added a small utility function `Config::ExpandDefaultConfigDir` that returns the expanded device client default config directory. I think @xlcheng1 is working on some additional use cases for pub-sub that require an expanded path. Please use this function.

**Review by HarshGandhi-AWS — APPROVED** [review]

**Review by xlcheng1 — APPROVED** [review]

**Pull request merged**